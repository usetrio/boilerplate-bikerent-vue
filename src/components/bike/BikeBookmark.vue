<script lang="ts">
import { Icon, IconType } from '@/components/icon'
import { defineComponent, type PropType } from 'vue'

export default defineComponent({
  name: 'BikeBookmark',
  components: {
    Icon
  },
  props: {
    active: {
      type: Boolean,
      default: false
    },
    outlined: {
      type: Boolean,
      default: false
    },
    size: {
      type: String as PropType<UISizing | UISizingType>,
      default: 'base'
    },
    width: {
      type: Number,
      default: 0
    }
  },
  emits: {
    'update:active': (_state: boolean) => true
  },
  computed: {
    iconType() {
      return this.active ? IconType.SOLID : IconType.REGULAR
    },
    componentClasses() {
      const classes: UICSSClass = ['bike-bookmark']

      classes.push({ 'bike-bookmark--outlined': this.outlined, 'bike-bookmark--active': this.active })

      return classes
    },
    componentStyle() {
      const style: UICSSProperties = {}

      if (this.width) {
        style.width = `${this.width}px`
      }

      return style
    }
  },
  methods: {
    handleToggleActive() {
      this.$emit('update:active', !this.active)
    }
  }
})
</script>

<template>
  <div :class="componentClasses" :style="componentStyle" @click.prevent="handleToggleActive">
    <icon :type="iconType" :size="size">heart</icon>
  </div>
</template>

<style lang="scss" scoped>
.bike-bookmark {
  position: relative;
  cursor: pointer;
  aspect-ratio: 1;

  @include modifier('outlined') {
    @apply .p-3, .inline-flex, .items-center, .justify-center;

    border: 1px solid get-theme-color('gray');
    border-radius: $border-radius-md;
  }
}
</style>
