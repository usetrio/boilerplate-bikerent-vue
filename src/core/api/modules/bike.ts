import type { BikeRent, BikeRentDetails, BikeReturnDetails } from './typings/bike'

// import { ApiResponseError, client } from '@/core/api'
import { client } from '@/core/api'

export type { BikeRent, BikeRentDetails, BikeReturnDetails }

/**
 * List Bikes
 *
 * @param {boolean} [available=false] List only available bikes for renting
 *
 * @returns {Promise<Bike[]>}
 */
export async function list(available = false): Promise<Bike[]> {
  return available ? await listAvailable() : await listAll()
}

/**
 * Lists all bikes.
 *
 * @returns {Promise<Bike[]>}
 */
export async function listAll(): Promise<Bike[]> {
  const { data } = await client.get<Bike[]>('bikes')

  return data
}

/**
 * Lists only available bikes, bikes that are not rented.
 *
 * @returns {Promise<Bike[]>}
 */
export async function listAvailable(): Promise<Bike[]> {
  const { data } = await client.get<Bike[]>('bikes/available')

  return data
}

/**
 * Rents a bike; you can also book future rents.
 *
 *
 * @param details
 * @returns
 */
export async function rent(details: BikeRentDetails): Promise<BikeRent> {
  const { data } = await client.post<BikeRent>('bikes/rent', details)

  return data
}

/**
 * Returns a bike
 * @param {BikeReturnDetails} details
 */
export async function returns(details: BikeReturnDetails): Promise<void> {
  await client.post('bikes/return', details)
}
