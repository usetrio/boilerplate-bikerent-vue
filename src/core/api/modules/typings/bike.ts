export interface BikeRent {
  rentAmount: number
  fee: number
  totalAmount: number
}

export interface BikeRentDetails {
  bikeId: number
  userId: number
  dateFrom: string
  dateTo: string
}

export interface BikeReturnDetails {
  bikeId: number
  userId: number
}
