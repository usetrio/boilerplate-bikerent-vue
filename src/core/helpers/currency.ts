/**
 * Formats value to currency notation
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat
 *
 * @param {number} value
 * @param {string} currency Currency Acronym (EUR, USD, BRL)
 * @returns {string}
 */
export function formatCurrency(value: number, currency: string): string {
  const formatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency,
    currencyDisplay: 'symbol',
    maximumFractionDigits: value % 1 > 0 ? 2 : 0
  })

  const parts = formatter.formatToParts(value)

  const currencyLabel = parts.find(({ type }) => type === 'currency')!.value
  const formattedValue = parts
    .filter(({ type }) => type !== 'currency')
    .map(({ value }) => value)
    .reduce((string, part) => string + part)

  return `${formattedValue} ${currencyLabel}`
}
