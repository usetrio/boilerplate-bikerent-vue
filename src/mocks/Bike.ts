import type { BikeRentDetails, BikeReturnDetails } from '@/core/api/modules/bike'

const [mockDate] = new Date().toISOString().split('T')

const mockedImageUrls = [
  'https://cremecycles.com/images/glowne/15.jpg',
  'https://cremecycles.com/images/glowne/13.jpg',
  'https://cremecycles.com/images/glowne/14.jpg'
]

const mockedBike: Bike = {
  id: 1,
  candidateId: 5,
  name: 'Murazik, Thiel and Robel',
  type: 'Cyclocross Bicycle',
  description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
  bodySize: 24,
  city: 'Manhattan',
  cardImage: 'https://cremecycles.com/images/glowne/15.jpg',
  imageUrls: mockedImageUrls,
  maxLoad: 110,
  rate: 135,
  ratings: 4.8
}

const mockBikeRentDetails: BikeRentDetails = {
  userId: 1,
  bikeId: mockedBike.id,
  dateFrom: mockDate,
  dateTo: mockDate
}

const mockedBikeReturnDetails: BikeReturnDetails = {
  userId: 1,
  bikeId: mockedBike.id
}

const mockedBikeList = [mockedBike, mockedBike, mockedBike, mockedBike]

export { mockedBike, mockedBikeList, mockedImageUrls, mockBikeRentDetails, mockedBikeReturnDetails }
