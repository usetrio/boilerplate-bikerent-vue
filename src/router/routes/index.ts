import HomeView from '@/views/HomeView.vue'
import NotFoundView from '@/views/Error/NotFoundView.vue'
import type { RouteRecordRaw } from 'vue-router'
import modules from './modules'

const routes: RouteRecordRaw[] = [
  {
    path: '/',
    name: 'home',
    component: HomeView
  },
  ...modules,
  {
    path: '/:pathMatch(.*)*',
    component: NotFoundView
  }
]

export default routes
