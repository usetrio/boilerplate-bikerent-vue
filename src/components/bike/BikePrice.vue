<script lang="ts">
import { defineComponent, type PropType } from 'vue'

import type { CurrencyCode } from '@/core/config'
import { formatCurrency } from '@/core/helpers/currency'

export type PropRateType = 'daily' | 'weekly' | 'monthly' | 'anually' | 'none'

export default defineComponent({
  name: 'BikePrice',
  props: {
    price: {
      type: Number,
      required: true
    },
    currency: {
      type: String as PropType<CurrencyCode>,
      required: true
    },
    rate: {
      type: String as PropType<PropRateType>,
      default: ''
    }
  },
  computed: {
    formattedPrice() {
      return formatCurrency(this.price, this.currency)
    },
    rateLabel() {
      let label = ''
      switch (this.rate) {
        case 'daily':
          label = 'Day'
          break
        case 'weekly':
          label = 'Week'
          break
        case 'monthly':
          label = 'Month'
          break
        case 'anually':
          label = 'Year'
          break
        default:
      }

      return label
    }
  }
})
</script>

<template>
  <div class="bike__price">
    <strong>{{ formattedPrice }}</strong>
    <template v-if="rateLabel">/ {{ rateLabel }}</template>
  </div>
</template>

<style lang="scss" scoped>
@include block('bike__price') {
  font-weight: 600;
  font-size: 16px;
  line-height: 23px;

  strong {
    font-weight: 800;
    font-size: 24px;
    line-height: 31px;
  }
}
</style>
