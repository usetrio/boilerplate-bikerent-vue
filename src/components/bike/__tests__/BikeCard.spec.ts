import { BikeBookmark, BikeCard, BikePrice } from '@/components/bike'
import { describe, expect, it, vi } from 'vitest'
import { flushPromises, mount } from '@vue/test-utils'

import { CurrencyCode } from '@/core/config'
import { mockIntersectionObserver } from '@/core/helpers/tests'
import { mockedBike } from '@/mocks/'
import router from '@/router'

describe('BikeCard', () => {
  vi.stubGlobal('IntersectionObserver', mockIntersectionObserver)

  it('renders correctly', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike }, global: { plugins: [router] } })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('displays title', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    expect(wrapper.find('.bike-card__title').text()).equals(mockedBike.name)
  })

  it('renders url', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })
    const { fullPath: routeStub } = router.resolve({ name: 'details', params: { id: mockedBike.id } })

    expect(wrapper.findComponent({ name: 'router-link' }).exists()).toBe(true)
    expect(wrapper.findComponent({ name: 'router-link' }).find('a').attributes('href')).equals(routeStub)
  })

  it('renders bookmark button', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    expect(wrapper.findComponent(BikeBookmark).exists()).toBe(true)
  })

  it('changes bookmark state', async () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    const bikeBookmark = wrapper.findComponent(BikeBookmark)

    await flushPromises()

    bikeBookmark.trigger('click')
    await flushPromises()

    expect(bikeBookmark.classes()).contain('bike-bookmark--active')
  })

  it('renders image', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    expect(wrapper.find('.bike-card__image img').exists()).toBe(true)
  })

  it('renders price correctly', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })
    const bikePriceStub = mount(BikePrice, {
      props: { price: mockedBike.rate, currency: CurrencyCode.EUR, rate: 'daily' }
    })

    expect(wrapper.findComponent(BikePrice).exists()).toBe(true)
    expect(wrapper.findComponent(BikePrice).text()).equal(bikePriceStub.text())
  })

  it('displays fluid layout', () => {
    const wrapper = mount(BikeCard, { props: { data: mockedBike, fluid: true }, global: { plugins: [router] } })

    expect(wrapper.classes()).includes('bike-card--fluid')
  })
})
