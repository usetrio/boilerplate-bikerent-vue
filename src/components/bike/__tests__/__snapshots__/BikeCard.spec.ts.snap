// Vitest Snapshot v1

exports[`BikeCard > renders correctly 1`] = `
"<div class=\\"bike-card\\" data-v-2b14ae81=\\"\\">
  <div class=\\"card\\" data-v-8955d826=\\"\\" data-v-2b14ae81=\\"\\">
    <div class=\\"card__image\\" data-v-8955d826=\\"\\">
      <div class=\\"bike-card__image\\" data-v-2b14ae81=\\"\\"><a href=\\"/details/1\\" class=\\"\\" data-v-2b14ae81=\\"\\"><img class=\\"image image--lazy\\" data-v-ff3f7011=\\"\\" data-v-2b14ae81=\\"\\"></a></div>
      <div class=\\"bike-card__bookmark\\" data-v-2b14ae81=\\"\\">
        <div class=\\"bike-bookmark\\" data-v-1f9ffc06=\\"\\" data-v-2b14ae81=\\"\\">
          <div class=\\"icon text-base\\" data-v-6ea13f32=\\"\\" data-v-1f9ffc06=\\"\\"><em class=\\"fa-regular fa-heart\\" data-v-6ea13f32=\\"\\"></em></div>
        </div>
      </div>
    </div>
    <div class=\\"card__title\\" data-v-8955d826=\\"\\">
      <div class=\\"bike-card__title\\" data-v-2b14ae81=\\"\\">
        <h3 data-v-2b14ae81=\\"\\"><a href=\\"/details/1\\" class=\\"\\" data-v-2b14ae81=\\"\\">Murazik, Thiel and Robel</a></h3>
      </div>
    </div>
    <div class=\\"card__content\\" data-v-8955d826=\\"\\">
      <div class=\\"flex items-center\\" data-v-2b14ae81=\\"\\">
        <div class=\\"chip chip--xs chip--secondary\\" data-v-221b4c58=\\"\\" data-v-2b14ae81=\\"\\">
          <div class=\\"chip__content\\" data-v-221b4c58=\\"\\">Cyclocross Bicycle</div>
        </div>
        <div class=\\"ml-auto\\" data-v-2b14ae81=\\"\\">
          <div class=\\"bike__price\\" data-v-14e6ff3d=\\"\\" data-v-2b14ae81=\\"\\"><strong data-v-14e6ff3d=\\"\\">135 â‚¬</strong>/ Day</div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
